Sources/Apollo/ApolloStore.swift

```diff
@@ -321,72 +261,25 @@ public final class ApolloStore {
                        withKey key: CacheKey,
                        variables: GraphQLMap?) throws {
       let normalizer = GraphQLResultNormalizer()
-      let executor = GraphQLExecutor(queue: queue) { object, info in
-        return .result(.success(object[info.responseKeyForField]))
+      let executor = GraphQLExecutor { object, info in
+        return object[info.responseKeyForField]
       }
-
+      
       executor.cacheKeyForObject = self.cacheKeyForObject
-
-      _ = try executor.execute(selections: selections,
-                               on: object,
-                               withKey: key,
-                               variables: variables,
-                               accumulator: normalizer)
-      .flatMap {
-        self.cache.mergePromise(records: $0)
-      }.andThen { changedKeys in
-        // Remove cached values from the data loader, so subsequent reads
-        // within the same transaction will reload the updated value.
-        self.loader.removeAll()
-        
-        if let didChangeKeysFunc = self.updateChangedKeysFunc {
-          didChangeKeysFunc(changedKeys, nil)
-        }
-      }.await()
-    }
-  }
-}
-
-internal extension NormalizedCache {
-  func loadRecordsPromise(forKeys keys: [CacheKey]) -> Promise<[Record?]> {
-    return Promise { fulfill, reject in
-      self.loadRecords(
-        forKeys: keys,
-        callbackQueue: nil) { result in
-          switch result {
-          case .success(let records):
-            fulfill(records)
-          case .failure(let error):
-            reject(error)
-          }
-        }
-    }
-  }
-
-  func mergePromise(records: RecordSet) -> Promise<Set<CacheKey>> {
-    return Promise { fulfill, reject in
-      self.merge(
-        records: records,
-        callbackQueue: nil) { result in
-          switch result {
-          case .success(let cacheKeys):
-            fulfill(cacheKeys)
-          case .failure(let error):
-            reject(error)
-          }
-      }
-    }
-  }
-
-  func clearPromise() -> Promise<Void> {
-    return Promise { fulfill, reject in
-      self.clear(callbackQueue: nil) { result in
-        switch result {
-        case .success(let success):
-          fulfill(success)
-        case .failure(let error):
-          reject(error)
-        }
+      
+      let records = try executor.execute(selections: selections,
+                                         on: object,
+                                         withKey: key,
+                                         variables: variables,
+                                         accumulator: normalizer)
+      let changedKeys = try self.cache.merge(records: records)
+      
+      // Remove cached records, so subsequent reads
+      // within the same transaction will reload the updated value.
+      loader.removeAll()
```

Comment:
why `removeAll` instead of the records that have been changed?

Reply:
I thought about it, and it might still be a good optimization to keep in mind, but because reads after writes within the same transaction are pretty uncommon, I didn't think it mattered that much. One problem is that our current cache keys don't differentiate between records and fields, so you'd be trying to remove a whole bunch of keys that didn't correspond to records after each write.

Reply:
Fair enough!